src/java.base/share/classes/javax/crypto/KEMSpi.java

```diff
@@ -0,0 +1,244 @@
+ * be immutable. It must be safe to invoke multiple {@code encapsulate} and
+ * {@code decapsulate} methods at the same time. Each invocation of
+ * {@code encapsulate} should generate a new shared secret and key
+ * encapsulation message.
+ * <p>
+ * For example,
+ * {@snippet lang = java:
+ * public static class MyKEMImpl implements KEMSpi {
+ *
+ *     @Override
+ *     public KEMSpi.EncapsulatorSpi engineNewEncapsulator(PublicKey pk,
+ *             AlgorithmParameterSpec spec, SecureRandom secureRandom)
+ *             throws InvalidAlgorithmParameterException, InvalidKeyException {
+ *         if (!checkPublicKey(pk)) {
+ *             throw new InvalidKeyException("unsupported key");
+ *         }
+ *         if (!checkParameters(spec)) {
+ *             throw new InvalidAlgorithmParameterException("unsupported params");
+ *         }
+ *         return new MyEncapsulator(pk, spec, secureRandom);
+ *     }
+ *
+ *     class MyEncapsulator implements KEMSpi.EncapsulatorSpi {
+ *         MyEncapsulator(PublicKey pk, AlgorithmParameterSpec spec, SecureRandom secureRandom){
+ *             this.spec = spec != null ? spec : getDefaultParameters();
+ *             this.secureRandom = secureRandom != null ? secureRandom : getDefaultSecureRandom();
+ *             this.pk = pk;
+ *         }
+ *
+ *         @Override
+ *         public KEM.Encapsulated encapsulate(int from, int to, String algorithm) {
+ *             // ...
+ *         }
+ *
+ *         // ...
+ *     }
+ *
+ *     // ...
+ * }
+ * }
+ *
+ * @see KEM
+ * @since 21
+ */
+public interface KEMSpi {
+
+    /**
+     * The KEM encapsulator implementation, generated by
+     * {@link #engineNewEncapsulator} on the KEM sender side.
+     *
+     * @see KEM.Encapsulator
+     */
+    interface EncapsulatorSpi {
+        /**
+         * The key encapsulation function.
+         * <p>
+         * Each invocation of this method must generates a new secret key and key
+         * encapsulation message that is returned in an {@link KEM.Encapsulated} object.
+         * <p>
+         * An implementation must support the case where {@code from} is 0,
+         * {@code from} is the same as the output of {@code secretSize()},
+         * and {@code algorithm} is "Generic".
+         *
+         * @param from the initial index of the shared secret byte array
+         *          to be returned, inclusive
+         * @param to the final index of the shared secret byte array
+         *          to be returned, exclusive
+         * @param algorithm the algorithm name for the secret key that is returned
+         * @return an {@link KEM.Encapsulated} object containing a portion of
+         *          the shared secret as a key with the specified algorithm,
+         *          the key encapsulation message, and optional parameters.
+         * @throws ArrayIndexOutOfBoundsException if {@code from < 0},
+         *     {@code from > to}, or {@code to > secretSize()}
```

Comment:
Should this throw an `IndexOutOfBoundsException` instead here? That's what the DHKEM impl does and this seems more reasonable since you aren't actually trying to access an array.